### ISSUE TITLE
Performance regression with Ansible v2 using include + big hash map

### ISSUE THREAD

**Issue by Yannig** [ansible/ansible#12997]
##### Issue Type:

This is a bug report about speed regression between Ansible 1.9.4 and current devel version when combining big hash map and include statement.
##### Ansible Version:

```
ansible 2.0.0 (devel f528ad1848) last updated 2015/11/01 09:23:49 (GMT +200)
  lib/ansible/modules/core: (detached HEAD a3a7cdb9aa) last updated 2015/11/01 10:03:19 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD b1db0e893c) last updated 2015/11/01 10:03:19 (GMT +200)
```
##### Ansible Configuration:

Nothing special. The playbook work with  local connection.
##### Environment:

N/A
##### Summary:

I have speed regressions against various playbooks. With Ansible 1.9.4, this playbooks was completed in about 3 to 5s. With last version of Ansible, they take 7 more times.

All my tests use a local virtual machine.
##### Steps To Reproduce:

You can retrieve playbook that reproduce this issue at the following location: https://github.com/Yannig/ansible-issue-low-speed/

This repository have 2 playbooks: one using a big hash map (issue-vars.yml) but not the other (no-vars-mapping.yml).
##### Expected Results:

Actually, using ansible 1.9.4, the playbook take about 3 seconds:

```
time ansible-playbook issue-vars/issue-vars.yml 
```

``` yaml
PLAY [Test] ******************************************************************* 

TASK: [do-include | debug var=application_name] ******************************* 
ok: [localhost] => {
    "var": {
        "application_name": "application1"
    }
}

TASK: [do-include | Répertoire des roles] ************************************ 
ok: [localhost] => (item={'key': 'field1', 'value': {'field': u'NA', 'name': 'field1'}})

TASK: [do-include | template src=test.j2 dest="/tmp/{{item.value.field}}/1-{{item.value.name}}.txt"] *** 
ok: [localhost] => (item={'key': 'field1', 'value': {'field': u'NA', 'name': 'field1'}})

TASK: [do-include | shell export FIELD={{item.item.value.field}} ; echo "$FIELD" ; cat /tmp/{{item.item.value.field}}/{{item.item.value.name}}.txt
] *** 
[...]
TASK: [do-include | shell export FIELD={{item.item.value.field}} ; echo "$FIELD" ; cat /tmp/{{item.item.value.field}}/{{item.item.value.name}}.txt
] *** 
skipping: [localhost] => (item=oracle_element_to_update.results)

PLAY RECAP ******************************************************************** 
localhost                  : ok=10   changed=0    unreachable=0    failed=0   
```

```
real    0m2.760s
user    0m2.052s
sys     0m0.121s
```
##### Actual Results:

Ansible v2 is about 7 times slower:

```
time ansible-playbook issue-vars/issue-vars.yml 
```

``` yaml
PLAY [Test] ********************************************************************

TASK [do-include : debug var=application_name] *********************************
ok: [localhost] => {
    "application_name": "application1",                                                                                                                                                                                                      
    "changed": false                                                                                                                                                                                                                         
}                                                                                                                                                                                                                                            

TASK [do-include : Répertoire des roles] ***************************************
ok: [localhost] => (item={'value': {u'field': u'NA', u'name': u'field1'}, 'key': u'field1'})

TASK [do-include : include] ****************************************************
included: /home/yannig/dev/yannig/ansible-issue-low-speed/issue-vars/roles/do-include/tasks/include1.yml for localhost

TASK [do-include : template dest=/tmp/{{item.value.field}}/1-{{item.value.name}}.txt src=test.j2] ***
ok: [localhost] => (item={'value': {u'field': u'NA', u'name': u'field1'}, 'key': u'field1'})

TASK [do-include : include] ****************************************************
included: /home/yannig/dev/yannig/ansible-issue-low-speed/issue-vars/roles/do-include/tasks/include2.yml for localhost
[...]
TASK [do-include : command] ****************************************************
skipping: [localhost] => (item=oracle_element_to_update.results) 

PLAY RECAP *********************************************************************
localhost                  : ok=12   changed=0    unreachable=0    failed=0   

```

```
real    0m20.879s
user    0m17.448s
sys     0m0.317s
```

If I take the other playbook **no-vars-mapping.yml** (which simply does not call vars-mapper role), Ansible v2 takes 2.2 seconds (vs 0.5 second for Ansible 1.9.4):

```
real    0m2.263s
user    0m1.318s
sys     0m0.178s
```

**Comment by Yannig** [comment]
The problem seems to come from **_do_template** call. The following patch gives some enhancement using a hash map as a cache:

``` diff
diff --git a/lib/ansible/template/__init__.py b/lib/ansible/template/__init__.py
index b7a0ea7..165b9e0 100644
--- a/lib/ansible/template/__init__.py
+++ b/lib/ansible/template/__init__.py
@@ -122,6 +122,7 @@ class Templar:
         self._filters             = None
         self._tests               = None
         self._available_variables = variables
+        self._cached_result       = {}

         if loader:
             self._basedir = loader.get_basedir()
@@ -298,18 +299,24 @@ class Templar:
                             elif resolved_val is None:
                                 return C.DEFAULT_NULL_REPRESENTATION

-                    result = self._do_template(variable, preserve_trailing_newlines=preserve_trailing_newlines, escape_backslashes=escape_backslashes, fail_on_undefined=fail_on_undefined, overrides=overrides)
-
-                    if convert_data:
-                        # if this looks like a dictionary or list, convert it to such using the safe_eval method
-                        if (result.startswith("{") and not result.startswith(self.environment.variable_start_string)) or \
-                           result.startswith("[") or result in ("True", "False"):
-                            eval_results = safe_eval(result, locals=self._available_variables, include_exceptions=True)
-                            if eval_results[1] is None:
-                                result = eval_results[0]
-                            else:
-                                # FIXME: if the safe_eval raised an error, should we do something with it?
-                                pass
+                    # Using a cache in order to prevent template calls with already templated variables
+                    cache_key = variable + str(preserve_trailing_newlines) + str(escape_backslashes) + str(overrides)
+                    try:
+                        result = self._cached_result[cache_key]
+                    except KeyError:
+                        result = self._do_template(variable, preserve_trailing_newlines=preserve_trailing_newlines, escape_backslashes=escape_backslashes, fail_on_undefined=fail_on_undefined, overrides=overrides)
+                        if convert_data:
+                            # if this looks like a dictionary or list, convert it to such using the safe_eval method
+                            if (result.startswith("{") and not result.startswith(self.environment.variable_start_string)) or \
+                              result.startswith("[") or result in ("True", "False"):
+                                eval_results = safe_eval(result, locals=self._available_variables, include_exceptions=True)
+                                if eval_results[1] is None:
+                                    result = eval_results[0]
+                                else:
+                                    # FIXME: if the safe_eval raised an error, should we do something with it?
+                                    pass
+                        self._cached_result[cache_key] = result
+

                 #return self._clean_data(result)
                 return result
```

This patch result in a 4 seconds playbook run (vs 20-21 seconds without patch):

```
real    0m3.988s
user    0m2.815s
sys     0m0.206s
```

It's not as good as with 1.9.4 (3s), but it's pretty close.

Now, it may have drawback ...

**Comment by bcoca** [comment]
closing as dupe of #12239

**Comment by Yannig** [comment]
Hi @bcoca 

I think it's not the same issue here. This issue talks about a problem with big hash map which can cause a huge regression in time execution.

**Comment by bcoca** [comment]
i believe they are very much related as the original issue seems to worsen the more vars are defined.

**Comment by Yannig** [comment]
Thanks for the feedback, I will follow this issue more closely in this case :)

**Comment by ghost** [comment]
We also experienced a huge performance drop when upgrading Ansible 1.9.x to 2.1.0, which was related to large yml variable files included. In our case, the performance drain was caused by the `deepcopy` introduced here: https://github.com/ansible/ansible/commit/d9833f227fb016563c389edb9fe5857feb6d5405 (the code was later moved here: https://github.com/ansible/ansible/blob/devel/lib/ansible/parsing/dataloader.py#L120)

Removing the `deepcopy` improved performance by a factor of 10 (~4 minutes instead of ~37 minutes for five hosts). We're not happy about removing the `deepcopy` (as it breaks the fix for #11893), but snice we aren't using any constructs which cause the problems described there it's currently our best choice.
